library/src/main/scala/scala/scalajs/js/package.scala

```diff
@@ -158,4 +158,64 @@ package object js {
    */
   def dynamicImport[A](body: => A): js.Promise[A] =
     throw new java.lang.Error("stub")
+
+  /** <span class="badge badge-ecma2017" style="float: right;">ECMAScript 2017</span>
+   *  Executes a block of code under a JavaScript `async` context.
+   *
+   *  The block of code can await [[Promise js.Promise]]s using [[await js.await]].
+   *  Doing so will continue after the call to `js.await` when the given
+   *  `Promise` is resolved. If the `Promise` is rejected, the exception gets
+   *  rethrown at the call site.
+   *
+   *  A block such as `js.async { body }` is equivalent to an
+   *  immediately-applied JavaScript `async` function:
+   *
+   *  {{{
+   *  (async () => body)()
+   *  }}}
+   *
+   *  Example usage:
+   *
+   *  {{{
+   *  val p: js.Promise[String] = downloadSomething()
+   *  val result: js.Promise[Int] = js.async {
+   *    val text: String = js.await(p)
+   *    text.toInt
+   *  }
+   *  }}}
+   *
+   *  This method returns a [[Promise js.Promise]] that will be resolved with
+   *  the result of the code block. If the block throws an exception, the
+   *  `Promise` will be rejected.
+   *
+   *  Calls to `js.await` can only appear within a `js.async` block. They must
+   *  not be nested in any local method, class, by-name argument or closure.
+   *  The latter includes `for` comprehensions. They may appear within
+   *  conditional branches, `while` loops and `try/catch/finally` blocks.
+   *
+   *  <h2>Orphan `await`s in WebAssembly</h2>
+   *
+   *  When compiling for Scala.js-on-Wasm only, you can allow calls to
+   *  `js.await` anywhere, using the compiler option
+   *  `-P:scalajs:allowOrphanJSAwait`.
+   *
+   *  Calls to orphan `js.await`s are validated at run-time. There must exist
+   *  a dynamically enclosing `js.async { ... }` block on the call stack.
+   *  Moreover, there cannot be any JavaScript frame (function invocation) in
+   *  the call stack between the `js.async { ... }` block and the call to
```

Comment:
Just out of curiosity, I believe this restriction comes from the JSPI specification https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md#restriction

Does this happen, for example, the invocation chain is something like:

- `A` wrapped by `WebAssembly.promising` →
- (non-suspending) JS function `B` →
- Wasm function `C`→ 
- suspending JS function `D`

?

Reply:
Yes, that is exactly the kind of invocation chain that is restricted by JSPI.